### PR TITLE
feat: add failure classification to E2E tests

### PR DESCRIPTION
**Pull request by cgoodfred**
## Description

JIRA: [RHOAIENG-49726](https://issues.redhat.com/browse/RHOAIENG-49726) 
- Adds implementation of test failure classifier, bases classifications on information generated by the health check reports for a cluster
- Defines `FailureClassification` struct, error code ranges, pattern tables for matching that can be easily updated to acommodate more scenarios
- Provides the output in 2 ways, `EmitClassification` writes human readable output to `stderr` and the structured JSON that is intended to be read by `test-retry`
- Updated logging in `debug_utils_test.go` to be based on the cluster health report but maintain similar behavior to what was previously logged

## Additional Notes
- Possibly replace the updated logging in `debug_utils_test.go` with `PrettyPrint` line once this [PR is merged](https://github.com/opendatahub-io/opendatahub-operator/pull/3216)
- Currently failure classification for genuine test failures is generic, currently investigating if its possible to expand on test failure categories further. Potentially for a follow-on PR.

## How Has This Been Tested?

I created a go file with some examples of report scenarios and ran it using the `EmitClassification` function to generate the output. The separator contains the scenario name and the produced. The actual functionality depends on the open PR to generate the content of the cluster health report.

```
=== clean cluster (all healthy) ===
2026/03/05 08:42:22 Classification for Test/clean cluster (all healthy): test/test-failure (code=2001, confidence=medium)
2026/03/05 08:42:22   Evidence: cluster state appears healthy, failure is likely test-related
FAILURE_CLASSIFICATION: {"category":"test","subcategory":"test-failure","error_code":2001,"evidence":["cluster state appears healthy, failure is likely test-related"],"confidence":"medium"}

=== ImagePullBackOff ===
2026/03/05 08:42:22 Classification for Test/ImagePullBackOff: infrastructure/image-pull (code=1001, confidence=medium)
2026/03/05 08:42:22   Evidence: container odh-dashboard-abc123/dashboard waiting: Back-off pulling image "quay.io/opendatahub/dashboard:bad-tag" ImagePullBackOff
FAILURE_CLASSIFICATION: {"category":"infrastructure","subcategory":"image-pull","error_code":1001,"evidence":["container odh-dashboard-abc123/dashboard waiting: Back-off pulling image \"quay.io/opendatahub/dashboard:bad-tag\" ImagePullBackOff"],"confidence":"medium"}

=== CrashLoopBackOff ===
2026/03/05 08:42:22 Classification for Test/CrashLoopBackOff: infrastructure/pod-startup (code=1002, confidence=medium)
2026/03/05 08:42:22   Evidence: container model-controller-xyz789/manager waiting: CrashLoopBackOff: back-off 5m0s restarting failed container
FAILURE_CLASSIFICATION: {"category":"infrastructure","subcategory":"pod-startup","error_code":1002,"evidence":["container model-controller-xyz789/manager waiting: CrashLoopBackOff: back-off 5m0s restarting failed container"],"confidence":"medium"}
...
```

I also ran some E2E tests on a cluster and had claude help force a failure to see the actual output and classification, some logs removed because there were many recent events. The test name is currently listed as components, but that isn't changed behavior from main, investigating if the test name can be more specific.
```
2026/03/08 22:04:25 TEST FAILURE DETECTED: components
2026/03/08 22:04:25 === RUNNING CLUSTER DIAGNOSTICS (triggered by test:components) ===
2026/03/08 22:04:25 === CLUSTER STATE ===
2026/03/08 22:04:25 Node: <ip>.ec2.internal
2026/03/08 22:04:25   Condition MemoryPressure: False - kubelet has sufficient memory available
2026/03/08 22:04:25   Condition DiskPressure: False - kubelet has no disk pressure
2026/03/08 22:04:25   Condition PIDPressure: False - kubelet has sufficient PID available
2026/03/08 22:04:25   Condition Ready: True - kubelet is posting ready status
2026/03/08 22:04:25   Resources: 7500m CPU, 31236232Ki memory allocatable (of 8000m CPU, 32387208Ki memory capacity)
2026/03/08 22:04:25 Node: <ip>.ec2.internal
2026/03/08 22:04:25   Condition MemoryPressure: False - kubelet has sufficient memory available
2026/03/08 22:04:25   Condition DiskPressure: False - kubelet has no disk pressure
2026/03/08 22:04:25   Condition PIDPressure: False - kubelet has sufficient PID available
2026/03/08 22:04:25   Condition Ready: True - kubelet is posting ready status
2026/03/08 22:04:25   Resources: 7500m CPU, 30892164Ki memory allocatable (of 8000m CPU, 32043140Ki memory capacity)
2026/03/08 22:04:25 === DEPLOYMENTS ===
2026/03/08 22:04:25 Failed to collect deployment data: deployments not ready: opendatahub/kuberay-operator (0/1)
2026/03/08 22:04:25 === PODS ===
2026/03/08 22:04:25 Failed to collect pod data: opendatahub/kuberay-operator-5c76565c5d-x9b45: phase=Pending
2026/03/08 22:04:25 === OPERATOR STATUS ===
2026/03/08 22:04:25 Operator deployment: 3/3 ready
2026/03/08 22:04:25 === DSCI STATUS ===
2026/03/08 22:04:25 DSCI default-dsci:
2026/03/08 22:04:25   ReconcileComplete: True - Reconcile completed successfully
2026/03/08 22:04:25   Available: True - Reconcile completed successfully
2026/03/08 22:04:25   Progressing: False - Reconcile completed successfully
2026/03/08 22:04:25   Degraded: False - Reconcile completed successfully
2026/03/08 22:04:25   Upgradeable: True - Reconcile completed successfully
2026/03/08 22:04:25 === DSC STATUS ===
2026/03/08 22:04:25 Failed to collect DSC data: default-dsc unhealthy conditions: Ready=False; ComponentsReady=False; RayReady=False
2026/03/08 22:04:25 === RECENT EVENTS ===
2026/03/08 22:04:25   22:04:14 WARN Pod/odh-dashboard-7d9f59b597-bnzvw: Unhealthy - Readiness probe failed: Get "https://10.129.0.175:8444/healthz": dial tcp 10.129.0.175:8444: connect: connection refused
...
2026/03/08 22:04:25   21:59:28 INFO Deployment/odh-model-controller: ScalingReplicaSet - Scaled down replica set odh-model-controller-f8f8cc8b from 1 to 0
2026/03/08 22:04:25 === RESOURCE QUOTAS ===
2026/03/08 22:04:25 No resource quotas found in monitored namespaces
2026/03/08 22:04:25 Classification for components: infrastructure/storage (code=1006, confidence=medium)
2026/03/08 22:04:25   Evidence: event Pod/kserve-controller-manager-789c69f8c4-w85p5: FailedMount - MountVolume.SetUp failed for volume "cert" : secret "kserve-webhook-server-cert" not found
FAILURE_CLASSIFICATION: {"category":"infrastructure","subcategory":"storage","error_code":1006,"evidence":["event Pod/kserve-controller-manager-789c69f8c4-w85p5: FailedMount - MountVolume.SetUp failed for volume \"cert\" : secret \"kserve-webhook-server-cert\" not found"],"confidence":"medium"}
2026/03/08 22:04:25 Diagnostics complete for failed test: components
    controller_test.go:317: Stopping: components test failed.
```

## Merge criteria

- [x] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [x] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [x] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work
- [x] The developer has run the integration test pipeline and verified that it passed successfully

### E2E test suite update requirement

When bringing new changes to the operator code, such changes are by default required to be accompanied by extending and/or updating the E2E test suite accordingly.

To opt-out of this requirement:
1. **Please inspect the [opt-out guidelines](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/e2e-update-requirement-guidelines.md)**, to determine if the nature of the PR changes allows for skipping this requirement
2. If opt-out is applicable, provide justification in the dedicated `E2E update requirement opt-out justification` section below
3. Check the checkbox below:
- [ ] Skip requirement to update E2E test suite for this PR
4. Submit/save these changes to the PR description. This will automatically trigger the check.

#### E2E update requirement opt-out justification

Assisted-By: Claude Opus 4.6 <noreply@anthropic.com>

## Summary by CodeRabbit

* **New Features**
  * Intelligent failure classification (infrastructure, test, unknown) with confidence, subcategories, and error codes.
  * Structured diagnostics orchestration: collects cluster health on panics/tests, emits human-readable summaries and machine-readable classification lines, and caches the latest classification for CI visibility.
  * Output helpers: formatted JUnit-style prefixes and a prefixed JSON line for emitted classifications.

* **Tests**
  * Extensive tests validating classification rules, edge cases, output formatting, and JSON round-trips.

### CODE REVIEW

File: tests/e2e/pkg/failureclassifier/output_test.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+	// Capture stdout
+	oldStdout := os.Stdout
+	r, w, _ := os.Pipe()
+	os.Stdout = w
+
+	EmitClassification(fc, "TestSomething")
+
+	w.Close()
+	os.Stdout = oldStdout
```
[the remark this answers is not in the document]
Comment (cgoodfred) [reply tests/e2e/pkg/failureclassifier/output_test.go:33]: This has been addressed in the latest commit.

File: tests/e2e/debug_utils_test.go [hunk 2]
```diff
@@ -121,313 +122,261 @@ func runDiagnosticsOnce(key string) {
...
+		DSCI: types.NamespacedName{Name: dsciInstanceName},
+		DSC:  types.NamespacedName{Name: dscInstanceName},
+	}
+
+	report, err := clusterhealth.Run(context.TODO(), cfg)
 	if err != nil {
-		return nil, err
+		log.Printf("ERROR: Failed to collect diagnostics: %v", err)
+		return
 	}
```
[the remark this answers is not in the document]
Comment (cgoodfred) [reply tests/e2e/debug_utils_test.go]: This has been addressed in the latest commit, it was updated to a 2 minute timeout.

File: tests/e2e/pkg/failureclassifier/classifier_test.go [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+		},
+		{
+			name: "all sections errored returns unknown",
+			report: &clusterhealth.Report{
+				Pods:        clusterhealth.SectionResult[clusterhealth.PodsSection]{Error: "fail"},
+				Events:      clusterhealth.SectionResult[clusterhealth.EventsSection]{Error: "fail"},
+				Quotas:      clusterhealth.SectionResult[clusterhealth.QuotasSection]{Error: "fail"},
+				Nodes:       clusterhealth.SectionResult[clusterhealth.NodesSection]{Error: "fail"},
+				Deployments: clusterhealth.SectionResult[clusterhealth.DeploymentsSection]{Error: "fail"},
+			},
```
[the remark this answers is not in the document]
Comment (cgoodfred) [reply tests/e2e/pkg/failureclassifier/classifier_test.go:67]: this was addressed in the latest commit.

File: tests/e2e/debug_utils_test.go [hunk 4]
```diff
@@ -121,313 +122,261 @@ func runDiagnosticsOnce(key string) {
...
 
-	// Try common label patterns for deployment pods
-	selectors := []client.MatchingLabels{
-		{"app": deploymentName},
-		{"app.kubernetes.io/name": deploymentName},
-		{"deployment": deploymentName},
+	// Retrieve actual container logs for problematic pods using raw Pod objects
+	if len(report.Pods.Data.Data) > 0 {
+		logProblematicContainerLogs(report.Pods.Data.Data)
 	}
```
[the remark this answers is not in the document]
Comment (cgoodfred) [reply tests/e2e/debug_utils_test.go:276]: I think this is fine, the `logProblematicContainerLogs` function calls `logContainerLogs` which runs through `redactSensitiveInfo` so it should be covered.

File: tests/e2e/debug_utils_test.go [hunk 5]
```diff
@@ -549,232 +511,20 @@ func retrievePodLogs(namespace, podName, containerName string, previous bool) (s
...
-}
-
-func getDSCI() *unstructured.Unstructured {
+func getDSCI(ctx context.Context) *unstructured.Unstructured {
 	dsci := &unstructured.Unstructured{}
 	dsci.SetGroupVersionKind(gvk.DSCInitialization)
-	err := globalDebugClient.Get(context.TODO(), types.NamespacedName{Name: dsciInstanceName}, dsci)
-	if err != nil && !errors.IsNotFound(err) {
+	err := globalDebugClient.Get(ctx, types.NamespacedName{Name: dsciInstanceName}, dsci)
+	if err != nil {
```
Comment (StevenTobin) [inline tests/e2e/debug_utils_test.go]: We should still check `&& !errors.IsNotFound(err)` here so we're differentiating between not found and there was some error.

File: tests/e2e/pkg/failureclassifier/classifier.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+// Classify inspects a clusterhealth Report and categorizes the failure.
+// If report is nil, returns unknown/unclassifiable/3000/low.
+func Classify(report *clusterhealth.Report) FailureClassification {
+	if report == nil {
+		return unknown()
+	}
+
+	// Per-section classifiers, ordered by priority (first match wins).
+	sectionClassifiers := []func(*clusterhealth.Report) *FailureClassification{
```
Comment (StevenTobin) [inline tests/e2e/pkg/failureclassifier/classifier.go:33]: Why are we not including the DSC, DSCI or operator sections?

File: tests/e2e/pkg/classifier/classifier.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+						Confidence:  ConfidenceMedium,
+					}
+				}
+			}
+			if pod.Phase == "Pending" {
+				return &FailureClassification{
+					Category:    CategoryInfrastructure,
+					Subcategory: "pod-startup",
+					ErrorCode:   CodePodStartup,
+					Evidence:    []string{fmt.Sprintf("pod %s stuck in Pending", pod.Name)},
```
Comment (StevenTobin) [inline tests/e2e/pkg/classifier/classifier.go]: This says pod is stuck pending, but that may not be true. The pod could be in pending state for 1 second and then healthy the next but if the report is gathered at the wrong time we will report it as stuck pending.

We should probably have a configurable `stuck pending period` and check that against the age of the pod.

File: tests/e2e/pkg/classifier/classifier.go [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+					if c.Waiting != "" {
+						return &FailureClassification{
+							Category:    CategoryInfrastructure,
+							Subcategory: "cluster-distress",
+							ErrorCode:   CodeInfraUnknown,
+							Evidence:    []string{fmt.Sprintf("container %s/%s in unrecognized waiting state: %s", pod.Name, c.Name, c.Waiting)},
+							Confidence:  ConfidenceLow,
+						}
+					}
+					if c.Terminated != "" {
```
Comment (StevenTobin) [inline tests/e2e/pkg/classifier/classifier.go]: I think a pod that terminates, but terminates successfully, ie, an init pod that is in completed state will show up here erroneously. We would need to check the message or exit code.

File: tests/e2e/pkg/classifier/types.go [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+	{"InvalidImageName", "image-pull", CodeImagePull},
+	{"pulling image", "image-pull", CodeImagePull},
+	{"CrashLoopBackOff", "pod-startup", CodePodStartup},
+	{"CreateContainerConfigError", "pod-startup", CodePodStartup},
+	{"CreateContainerError", "pod-startup", CodePodStartup},
+}
+
+// Patterns matched against container Terminated message text (substring match).
+var terminatedPatterns = []classificationPattern{
+	{"OOMKilled", "quota-oom", CodeQuotaOOM},
```
Comment (StevenTobin) [inline tests/e2e/pkg/classifier/types.go]: OOMKilled because the pod limits/requests are too small should be a different failure to killed because of a quota in the namespace.

File: tests/e2e/debug_utils_test.go (hunk 5 above)
Comment (cgoodfred) [reply tests/e2e/debug_utils_test.go]: 👍 Added that back

File: tests/e2e/pkg/failureclassifier/classifier.go (hunk 6 above)
Comment (cgoodfred) [reply tests/e2e/pkg/failureclassifier/classifier.go:33]: Thats a good callout, I added sections for these. I'm a little unsure on the correct order here as a pod issue and operator issue are not mutually exclusive. At the moment the pod classification would occur first, thoughts on order?

File: tests/e2e/pkg/classifier/classifier.go (hunk 7 above)
Comment (cgoodfred) [reply tests/e2e/pkg/classifier/classifier.go]: I added a `CreatedAt` field in pod info and then defined a constant (currently set to 1 minute) to give it a waiting period based on the pending threshold and time since being created. From what I've tested locally this seems to work out but I'll verify tomorrow when my cluster is back up. LMK what you think

File: tests/e2e/pkg/classifier/classifier.go (hunk 8 above)
Comment (cgoodfred) [reply tests/e2e/pkg/classifier/classifier.go]: Good catch, I added a check to determine if the termination was for an exit code of 0 based on how the cluster health package includes the code and not include in that case.

File: tests/e2e/pkg/classifier/types.go (hunk 9 above)
Comment (cgoodfred) [reply tests/e2e/pkg/classifier/types.go]: I separated OOMKilled on the container terminated pattern to be its own code and left the `classifyFromQuotas` handling of the quota oom identifier.

File: tests/e2e/pkg/classifier/classifier.go [hunk 10]
```diff
@@ -0,0 +1,279 @@
+package classifier
```
Comment (macgregor) [inline tests/e2e/pkg/classifier/classifier.go]: nit: worth a separate package or should we just keep it in the clusterhealth pkg? 

regardless, verging on not a nit: classifier -> failureclassifier (same for func/types/wahtever, makes it more clear the purpose at a glance)

File: tests/e2e/pkg/failureclassifier/classifier.go [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+	return unknown()
+}
+
+
+// classifyFromPods checks container states and pod phases.
+// Covers: image-pull, pod-startup, OOM subcategories.
+func classifyFromPods(report *clusterhealth.Report) *FailureClassification {
+	for _, pods := range report.Pods.Data.ByNamespace {
+		for _, pod := range pods {
+			for _, container := range pod.Containers {
```
Comment (macgregor) [inline tests/e2e/pkg/failureclassifier/classifier.go:78]: triple nested loop. what sorts of list sizes can we expect here? once might not be so bad, but i see basically the same loop in classifyClusterDistress. so we're looking at a nasty number of permutations. still might be ok, but something to consider and look for maybe a better way to structure things to loop over things once and check everything you need to check while youre here. So like one big control loop basically, if possible, and you call some specific classifier function(s) at each level as needed.

File: tests/e2e/pkg/failureclassifier/classifier.go (hunk 6 above)
Comment (StevenTobin) [reply tests/e2e/pkg/failureclassifier/classifier.go:33]: I think pod is fine first, it's much more broad since it covers all pods and not just the operator. It makes sense to start broad and then get more specific later.

File: tests/e2e/pkg/classifier/classifier.go (hunk 10 above)
Comment (cgoodfred) [reply tests/e2e/pkg/classifier/classifier.go]: I went ahead and renamed the package to `failureclassifier` good call out on classifier being too generic. I think its useful to keep it separate from clusterhealth, it gives a clear distinction between the two since the clusterhealth reports are useful beyond just the failure classifier.

File: tests/e2e/pkg/failureclassifier/classifier.go (hunk 11 above)
Comment (cgoodfred) [reply tests/e2e/pkg/failureclassifier/classifier.go:78]: I think the number of permutations here won't be unreasonable to go through once, we're looking at a handful of namespaces with tens of pods and a few containers each and a very small footprint per container in the `ContainerInfo` struct. 

I do think we can avoid looping twice though, I added a second return to `classifyFromPods` that will catch the first signal that would have indicated pod distress and return that if we don't come across a categorized pod signal. Then pass it in to the generic cluster distress function, that way we still use known signals from pods or the other sections before we fall back on the unknown pod failure.

**Comment by kahowell** [comment]
/override "Red Hat Konflux / odh-operator-ci-e2e-its-manual / odh-operator-ci"
